core/src/main/java/io/fabric8/maven/core/openshift/BuildService.java

```diff
@@ -81,6 +81,15 @@ public void waitForOpenShiftBuildToComplete(OpenShiftClient client, Build build)
                                                   "Failed to tail build log", logTerminateLatch, log);
             Watcher<Build> buildWatcher = getBuildWatcher(latch, buildName, buildHolder);
             try (Watch watcher = client.builds().withName(buildName).watch(buildWatcher)) {
+                // Check if the build is already finished to avoid waiting indefinitely
+                Build lastBuild = client.builds().withName(buildName).get();
+                String lastStatus = KubernetesResourceUtil.getBuildStatusPhase(lastBuild);
+                if (Builds.isFinished(lastStatus)) {
+                    log.debug("Build %s is already finished", buildName);
+                    buildHolder.set(lastBuild);
+                    latch.countDown();
+                }
+
```

Comment:
No, it can't be moved. The build returned by Openshift is not "Complete" sometimes, and it is finished right after the response has been returned, so an additional fetch is needed after starting the watch, in order to check if the "Complete" event has been sent in the meantime (probably we should open an issue to Openshift, but version 3.6 is on the way).

Actually this issue is related to Openshift 3.6+ (not 3.7+, I was confused by last Kubernetes release number).

About the timeout, I just did a build that lasted 5+ minutes without issues. I think that this change (in Openshift) will not have impacts on client side timeouts.

At least, I've not experienced timeouts.